packages/gatsby-plugin-sitemap/src/gatsby-node.js

```diff
@@ -1,80 +1,139 @@
 import path from "path"
-import sitemap from "sitemap"
+import minimatch from "minimatch"
+import { simpleSitemapAndIndex } from "sitemap"
+import { validateOptions } from "./options-validation"
 import {
-  defaultOptions,
-  filterQuery,
-  writeFile,
-  renameFile,
   withoutTrailingSlash,
+  prefixPath,
+  defaultFilterPages,
+  defaultExcludes,
 } from "./internals"
 
 const publicPath = `./public`
 
+const reporterPrefix = `[plugin-sitemap]:`
+
+exports.onPreInit = async ({ reporter }, pluginOptions) => {
+  try {
+    await validateOptions(pluginOptions)
+    reporter.verbose(`${reporterPrefix} Plugin options passed validation.`)
+  } catch (err) {
+    reporter.panic(err)
+  }
+}
+
 exports.onPostBuild = async (
-  { graphql, pathPrefix, basePath = pathPrefix },
+  { graphql, reporter, pathPrefix },
   pluginOptions
 ) => {
-  const options = { ...pluginOptions }
-  delete options.plugins
-  delete options.createLinkInHead
+  global.reporter = reporter
 
   const {
-    query,
-    serialize,
     output,
-    exclude,
-    hostname,
+    entryLimit,
+    query,
+    excludes,
     resolveSiteUrl,
-    ...rest
-  } = {
-    ...defaultOptions,
-    ...options,
+    resolvePagePath,
+    resolvePages,
+    filterPages,
+    serialize,
+  } = await validateOptions(pluginOptions).catch(err => {
+    reporter.panic(err)
+  })
+
+  const { data: queryRecords } = await graphql(query)
+
+  reporter.verbose(
+    `${reporterPrefix} Query Results` + JSON.stringify(queryRecords, null, 2)
+  )
+
+  const allPages = resolvePages(queryRecords)
```

Comment:
Should we make it async, so users can do whatever? I don't see an immediate use case but 🤷

Reply:
I'll await `resolveSiteUrl` as well.